packages/client/src/rtc/Dispatcher.ts

```diff
@@ -1,8 +1,51 @@
-import { EventTypes, Logger } from '../coordinator/connection/types';
-import type { SfuEvent } from '../gen/video/sfu/event/events';
+import { CallEventListener, EventTypes } from '../coordinator/connection/types';
+import type {
+  AudioLevelChanged,
+  CallGrantsUpdated,
+  ChangePublishQuality,
+  ConnectionQualityChanged,
+  DominantSpeakerChanged,
+  Error as SfuError,
+  GoAway,
+  HealthCheckResponse,
+  ICERestart,
+  ICETrickle,
+  JoinResponse,
+  ParticipantJoined,
+  ParticipantLeft,
+  PinsChanged,
+  PublisherAnswer,
+  SfuEvent,
+  SubscriberOffer,
+  TrackPublished,
+  TrackUnpublished,
+} from '../gen/video/sfu/event/events';
 import { getLogger } from '../logger';
 
 export type SfuEventKinds = NonNullable<SfuEvent['eventPayload']['oneofKind']>;
+export type AllSfuEvents = {
+  // @ts-ignore - TS doesn't like this for some reason
+  // had to type it out manually :)
+  // [K in SfuEventKinds]: Extract<SfuEvent['eventPayload'], { oneofKind: K }>[K];
+  subscriberOffer: SubscriberOffer;
+  publisherAnswer: PublisherAnswer;
+  connectionQualityChanged: ConnectionQualityChanged;
+  audioLevelChanged: AudioLevelChanged;
+  iceTrickle: ICETrickle;
+  changePublishQuality: ChangePublishQuality;
+  participantJoined: ParticipantJoined;
+  participantLeft: ParticipantLeft;
+  dominantSpeakerChanged: DominantSpeakerChanged;
+  joinResponse: JoinResponse;
+  healthCheckResponse: HealthCheckResponse;
+  trackPublished: TrackPublished;
+  trackUnpublished: TrackUnpublished;
+  error: SfuError;
+  callGrantsUpdated: CallGrantsUpdated;
+  goAway: GoAway;
+  iceRestart: ICERestart;
+  pinsUpdated: PinsChanged;
+};
```

Comment:
TS just needs a little more assurance that K can be used to index the extracted type :)

```suggestion
export type AllSfuEvents = {
  [K in SfuEventKinds]: K extends keyof Extract<
    SfuEvent['eventPayload'],
    { oneofKind: K }
  >
    ? Extract<SfuEvent['eventPayload'], { oneofKind: K }>[K]
    : never;
};
```

Reply:
Nice! thank you for the optimization!